Added logic for working with Tarantool schema via Box

KaymeKaydex opened the pull request:
- Implemented the `box.Schema()` method that returns a `Schema` object for schema-related operations

What has been done? Why? What problem is being solved?

I didn't forget about (remove if it is not applicable):

- [x] Tests (see [documentation](https://pkg.go.dev/testing) for a testing package)
- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)
- [x] Documentation (see [documentation](https://go.dev/blog/godoc) for documentation style guide)

Related issues:

oleg-jukovec commented on box/schema_user.go:
```diff
@@ -0,0 +1,220 @@
+	conn tarantool.Doer // Connection interface for interacting with Tarantool.
+}
+
+// User returns a new SchemaUser instance, allowing schema-related user operations.
+func (s *Schema) User() *SchemaUser {
+	return &SchemaUser{conn: s.conn}
+}
```
The same about a constructor as for the `Schema` type.

KaymeKaydex replied:
done

oleg-jukovec commented on box/schema_user.go:
Please, add unit-test for requests and responses types in the file.

We probably also need to think about naming and location of requests types in this package. As example, at now some requests are at `request.go` file. It could be confusing.

There are two ways I see:
1. We could locate all requests/responses into the `request.go` file (I don't like the idea).
2. We need move types from `request.go` into a proper files in the same manner as for the current ones.

KaymeKaydex replied:
I propose to completely get rid of request.go since baseRequest is implemented incorrectly, the Info call is a simple CallRequest, which should inherit all the logic from a regular call request, as is done for schema user

KaymeKaydex replied:
All box requests are not custom requests with msgpack, thats why we need to use just simple call requests

KaymeKaydex commented on box/schema.go:
```diff
@@ -0,0 +1,17 @@
+	return &Schema{
+		conn: b.conn, // Pass the Box connection to the Schema.
+	}
+}
```
done

KaymeKaydex commented on box/box_test.go:
```diff
@@ -10,7 +10,6 @@ import (
 func TestNew(t *testing.T) {
 	// Create a box instance with a nil connection. This should lead to a panic later.
 	b := box.New(nil)
-
```
done

KaymeKaydex commented on box/schema.go:
```diff
@@ -0,0 +1,17 @@
+	return &Schema{
+		conn: b.conn, // Pass the Box connection to the Schema.
+	}
+}
```
done

KaymeKaydex commented on box/tarantool_test.go:
```diff
@@ -61,6 +63,85 @@ func TestBox_Info(t *testing.T) {
+	exists, err = b.Schema().User().Exists(ctx, username)
+	require.False(t, exists)
+	require.NoError(t, err)
+}
```
done

KaymeKaydex replied:
i wrote some sub-tests, that looks better

oleg-jukovec commented on CHANGELOG.md:
```diff
@@ -10,8 +10,12 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
+
 ### Changed
 
+- Box Info method now requires context and implements CallRequest type instead custom baseRequest.
```
This is a breaking change. We need an another method and keep a backward compatibility here. You could create an issue to delete an old method in the next major release.

KaymeKaydex replied:
Removed

oleg-jukovec reviewed (changes requested):
Please, add more unit-tests for a new requests/response. You could see how it is tested on the main package and inside the `crud` subpackage.

We also need examples (in example_test.go) with new features. This will help users to use the library.

Finally, we need to keep a backward compatibility until a next major release, so we need to add another Info() method with a context instead of changing the current one.

KaymeKaydex commented:
> Please, add more unit-tests for a new requests/response. You could see how it is tested on the main package and inside the `crud` subpackage.
> 
> We also need examples (in example_test.go) with new features. This will help users to use the library.
> 
> Finally, we need to keep a backward compatibility until a next major release, so we need to add another Info() method with a context instead of changing the current one.

i removed interface changes
added new examples for new methods
and added more tests to make coverage up, now all methods are covered

oleg-jukovec commented on CHANGELOG.md:
```diff
@@ -11,6 +11,8 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 - Extend box with replication information (#427).
+- Implemented all box.schema.user operations requests and sugar interface.
+- Implemented box.session.su request and sugar interface only for current session granting.
```
```suggestion
- Implemented box.session.su request and sugar interface only for current session granting (#426).
```

oleg-jukovec commented on box/example_test.go:
```diff
@@ -58,3 +58,129 @@ func Example() {
+		log.Fatalf("Box schema users exists are not equal")
+	}
+
+	fmt.Printf("Box schema users exists are equal")
+	fmt.Printf("Current exists state: %+v\n", exists)
```
Why all the examples are not runable as tests? You can run tarantula before running the example as it is done in other tests.

```suggestion
	fmt.Printf("Box schema users exists are equal")
	fmt.Printf("Current exists state: %+v\n", exists)

	// Output:
	// Box schema users exists are equal
	// Current exists state: true
```

KaymeKaydex replied:
fixed

oleg-jukovec commented on box/example_test.go:
```diff
@@ -58,3 +58,129 @@ func Example() {
+}
+
+func ExampleSchemaUser_Create() {
+	// Connect to Tarantool
```
You add the comment in some tests, but not to another. It is a little confusing.
```suggestion
	// Connect to Tarantool.
```

oleg-jukovec commented on box/example_test.go:
```diff
@@ -58,3 +58,129 @@ func Example() {
+		log.Fatalf("Failed to connect: %s", err)
+	}
+
+	// Create SchemaUser
+	schemaUser := box.NewSchemaUser(client)
+
+	// Create a new user
```
```suggestion
	// Create SchemaUser.
	schemaUser := box.NewSchemaUser(client)

	// Create a new user.
```

oleg-jukovec commented on box/example_test.go:
```diff
@@ -58,3 +58,129 @@ func Example() {
+}
+
+func ExampleSchemaUser_Drop() {
+	// Connect to Tarantool
```
The same about dots: a comment should end with a dot. Please fix all comments in the file.

KaymeKaydex replied:
fixed

oleg-jukovec commented on box/schema_test.go:
```diff
@@ -0,0 +1,16 @@
+
+	// Ensure the schema is not nil (which it shouldn't be), but this is not meaningful
+	// since we will panic when we call the schema methods with the nil connection.
+	require.NotNil(t, b)
```
```suggestion
	require.NotNil(t, b)
	require.Nil(t, b.User())
```

KaymeKaydex replied:
it might be not nil, cause we panics only on nil connection

KaymeKaydex replied:
i added test with info method call panics, cause conn is nil

i think resolved

oleg-jukovec commented on box/schema_test.go:
```diff
@@ -0,0 +1,16 @@
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewSchema(t *testing.T) {
```
You could probably add a unit-test that a Schema.User() returns a SchemaUser object with the given doer (with a mock doer). Something like:
```
doer := struct{
    tarantool.Doer
}{}
require.Equal(*NewSchema(doer).User(), *NewSchemUser(doer))
```

KaymeKaydex replied:
i added such test for schema 

fixed

oleg-jukovec commented on box/schema_user.go:
```diff
@@ -0,0 +1,542 @@
+}
+
+// PrivilegeObjectType is a struct based on privilege object types list
+// https://www.tarantool.io/en/doc/latest/admin/access_control/#all-object-types-and-permissions
+type PrivilegeObjectType string
+
+const (
+	// PrivilegeObjectTypeUniverse - privilege type based on universe.
+	// A database (box.schema) that contains database objects, including spaces,
+	// indexes, users, roles, sequences, and functions.
+	// Granting privileges to universe gives a user access to any object in the database.
+	PrivilegeObjectTypeUniverse PrivilegeObjectType = "universe"
```
Can we use this naming?

```suggestion
// https://www.tarantool.io/en/doc/latest/admin/access_control/#all-object-types-and-permissions
type PrivilegeType string

const (
	// PrivilegeUniverse - privilege type based on universe.
	// A database (box.schema) that contains database objects, including spaces,
	// indexes, users, roles, sequences, and functions.
	// Granting privileges to universe gives a user access to any object in the database.
	PrivilegeUniverse PrivilegeType = "universe"
```

KaymeKaydex replied:
fixed

oleg-jukovec commented on box/schema_user.go:
```diff
@@ -0,0 +1,542 @@
+type Privilege struct {
+	// Permissions is a list of privileges that apply to the privileges object type.
+	Permissions []Permission
+	// ObjectType - one of privilege object types (it might be space,function, etc.).
+	ObjectType PrivilegeObjectType
+	// ObjectName - can be the name of a function or space,
+	// and can also be empty in case of universe access
+	ObjectName string
```
And such...
```suggestion
	// Type - one of privilege object types (it might be space,function, etc.).
	Type PrivilegeObjectType
	// Name - can be the name of a function or space,
	// and can also be empty in case of universe access
	Name string
```

KaymeKaydex replied:
Naming based on tarantool doc https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_schema/user_grant/#lua-function.box.schema.user.grant

KaymeKaydex replied:
fixed

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+
+	b := box.New(conn)
+
+	// Create new user
```
Please, checkk all comments for dots )
```suggestion
	// Create new user.
```

oleg-jukovec reviewed (commented):
Please, fix the commit message:


```
box: ddded logic for working with Tarantool schema

Implemented the `box.Schema()` method that returns a `Schema`
object for schema-related operations
```

KaymeKaydex commented:
> Please, fix the commit message:
> 
> ```
> box: ddded logic for working with Tarantool schema
> 
> Implemented the `box.Schema()` method that returns a `Schema`
> object for schema-related operations
> ```

done

oleg-jukovec commented on box/schema_test.go:
```diff
@@ -0,0 +1,45 @@
+
+	// Ensure the schema is not nil (which it shouldn't be), but this is not meaningful
+	// since we will panic when we call the schema methods with the nil connection.
+	t.Run("internal sugar sub-objects not panics", func(t *testing.T) {
+		require.NotNil(t, b)
+		require.NotNil(t, b.User())
+	})
+
+	t.Run("check that connections are equal", func(t *testing.T) {
+		var tCases []tarantool.Doer
+		for i := 0; i < 10; i++ {
+
+			doer := test_helpers.NewMockDoer(t)
+			tCases = append(tCases, &doer)
+		}
+
+		for _, tCase := range tCases {
+			sch := NewSchema(tCase)
+			require.Equal(t, tCase, sch.conn)
+			require.Equal(t, tCase, sch.User().conn)
+		}
+	})
+
+	t.Run("nil conn panics", func(t *testing.T) {
```
Please, split it into 3 separate tests.

KaymeKaydex replied:
done

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+	err = b.Schema().User().Create(ctx, username, box.UserCreateOptions{Password: password})
+	require.NoError(t, err)
+
+	t.Run("can connect with new credentials", func(t *testing.T) {
+		t.Parallel()
+		// Check that password is valid and we can connect to tarantool with such credentials
+		var newUserDialer = tarantool.NetDialer{
+			Address:  server,
+			User:     username,
+			Password: password,
+		}
+
+		// We can connect with our new credentials
+		newUserConn, err := tarantool.Connect(ctx, newUserDialer, tarantool.Opts{})
+		require.NoError(t, err)
+		require.NotNil(t, newUserConn)
+		require.NoError(t, newUserConn.Close())
+	})
+	t.Run("create user already exists error", func(t *testing.T) {
```
The same, please, split it into separate tests instead of the nesting.

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+
+	b := box.New(conn)
+
+	t.Run("drop user after create", func(t *testing.T) {
```
etc

KaymeKaydex replied:
done

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+
+	b := box.New(conn)
+
+	t.Run("user not found", func(t *testing.T) {
```
etc

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+
+	b := box.New(conn)
+
+	t.Run("test user privileges is correct", func(t *testing.T) {
```
etc

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+	err = b.Schema().User().Create(ctx, username, box.UserCreateOptions{Password: password})
+	require.NoError(t, err)
+
+	t.Run("can`t grant without su permissions", func(t *testing.T) {
```
etc

KaymeKaydex replied:
done

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+	err = b.Schema().User().Create(ctx, username, box.UserCreateOptions{Password: password})
+	require.NoError(t, err)
+
+	t.Run("can`t revoke without su permissions", func(t *testing.T) {
```
etc

KaymeKaydex replied:
done

oleg-jukovec commented on box/tarantool_test.go:
```diff
@@ -67,6 +69,407 @@ func TestBox_Info(t *testing.T) {
+
+	b := box.New(conn)
+
+	t.Run("su with call", func(t *testing.T) {
```
etc

KaymeKaydex replied:
done

KaymeKaydex commented:
@oleg-jukovec @DerekBum @themilchenko up

oleg-jukovec commented on box/box.go:
```diff
@@ -17,6 +17,12 @@ func New(conn tarantool.Doer) *Box {
 	}
 }
 
+// Schema returns a new Schema instance, providing access to schema-related operations.
+// It uses the connection from the Box instance to communicate with Tarantool.
+func (b *Box) Schema() *Schema {
+	return NewSchema(b.conn)
+}
```
It would be nice to write a test that checks whether the created circuit uses the connection passed to `box.New`.

oleg-jukovec reviewed (approved):
Thanks for the patch. It would be nice if @dmyger paid attention to the test coverage.

oleg-jukovec reviewed (changes requested):
See comments above.

oleg-jukovec commented:
Thank you for the PR! We've added some minor fixes and merged it: https://github.com/tarantool/go-tarantool/pull/446
